src/mlpack/methods/kmeans/elkan_kmeans_impl.hpp

```diff
@@ -78,15 +71,49 @@ double ElkanKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
   // that this is equivalent to s(c) for each cluster c.
   minClusterDistances = 0.5 * min(clusterDistances).t();
 
+  // If this is the first iteration, we must reset all the bounds.
+  if (lowerBounds.n_rows != centroids.n_cols)
+  {
+    lowerBounds.set_size(centroids.n_cols, dataset.n_cols);
+    assignments.set_size(dataset.n_cols);
+    upperBounds.set_size(dataset.n_cols);
+
+    lowerBounds.fill(0);
+    upperBounds.fill(DBL_MAX);
+    assignments.fill(0);
+  }
+
+  // Determine the number of threads
+  int numThreads = 1;
+  #ifdef MLPACK_USE_OPENMP
+    #pragma omp parallel
+    {
+      #pragma omp single
+      numThreads = omp_get_num_threads();
+    }
+  #endif
```

Comment:
When I take a look at the naive strategy, the idea there is to just create a local centroid object inside each thread. If you applied that strategy here too (and just used a `#pragma omp critical` when summing centroids) I think the code would be much simpler and you wouldn't need `numThreads`.  Let me know what you think.